### PR TITLE
 Modified code to support python 2.

### PR DESCRIPTION
**Pull request by rvelaVenafi**

### CODE REVIEW

File: tests/test_e2e.py [hunk 1]
```diff
@@ -40,7 +38,7 @@
 from cryptography.hazmat.primitives import serialization, hashes
 logging.basicConfig(level=logging.DEBUG)
 logging.getLogger("urllib3").setLevel(logging.DEBUG)
-
+from assets import *
```
Comment (warrior-abhijit) [inline tests/test_e2e.py:41]: replace with individual imports if possible

File: vcert/connection_tpp.py [hunk 2]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+            if request_data.get("CASpecificAttributes"):
+                request_data["CASpecificAttributes"].append(ca_origin)
+            else:
+                request_data["CASpecificAttributes"] = [ca_origin]
+        status, data = self._post(URLS.CERTIFICATE_REQUESTS, data=request_data)
+        if status == HTTPStatus.OK:
+            request.id = data['CertificateDN']
+            log.debug("Certificate sucessfully requested with request id %s." % request.id)
+            return True
+        else:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:152]: redundant else

File: vcert/connection_tpp.py [hunk 3]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+
+        status, data = self._post(URLS.CERTIFICATE_RETRIEVE, data=retrive_request)
+        if status == HTTPStatus.OK:
+            pem64 = data['CertificateData']
+            pem = base64.b64decode(pem64)
+            return parse_pem(pem.decode(), certificate_request.chain_option)
+        elif status == HTTPStatus.ACCEPTED:
+            log.debug(data['Status'])
+            return None
+        else:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:181]: redundant else

File: vcert/connection_tpp.py [hunk 4]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+        elif request.thumbprint:
+            d["Thumbprint"] = request.thumbprint
+        else:
+            raise ClientBadData
+        if request.comments:
+            d["Comments"] = request.comments
+        status, data = self._post(URLS.CERTIFICATE_REVOKE, data=d)
+        if status in (HTTPStatus.OK, HTTPStatus.ACCEPTED):
+            return data
+        else:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:204]: redundant else

File: vcert/connection_tpp.py [hunk 5]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+        if reuse_key:
+            log.debug("Trying to renew certificate %s" % request.id)
+            status, data = self._post(URLS.CERTIFICATE_RENEW, data={"CertificateDN": request.id})
+            if not data['Success']:
+                raise CertificateRenewError
+            return
+        cert = self.retrieve_cert(request)
+        cert = x509.load_pem_x509_certificate(cert.cert.encode(), default_backend())
+        for a in cert.subject:
+            if a.oid == x509.NameOID.COMMON_NAME:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:222]: switch case will be better here ?

File: vcert/connection_tpp.py [hunk 6]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+        if not request.csr:
+            request.build_csr()
+        status, data = self._post(URLS.CERTIFICATE_RENEW,
+                                  data={"CertificateDN": request.id, "PKCS10": request.csr})
+        if status == HTTPStatus.OK:
+            if "CertificateDN" in data:
+                request.id = data['CertificateDN']
+            log.debug("Certificate successfully requested with request id %s." % request.id)
+            return True
+        else:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:253]: redundant else

File: vcert/connection_tpp.py [hunk 7]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+        if cert.signature_algorithm_oid in (algos.ECDSA_WITH_SHA1, algos.ECDSA_WITH_SHA224, algos.ECDSA_WITH_SHA256,
+                                            algos.ECDSA_WITH_SHA384, algos.ECDSA_WITH_SHA512):
+            request.key_type = (KeyType.ECDSA, KeyType.ALLOWED_CURVES[0])
+        else:
+            request.key_type = KeyType(KeyType.RSA, 2048)  # todo: make parsing key size
+        if not request.csr:
+            request.build_csr()
+        status, data = self._post(URLS.CERTIFICATE_RENEW,
+                                  data={"CertificateDN": request.id, "PKCS10": request.csr})
+        if status == HTTPStatus.OK:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:248]: how about combining two if's into one if statement with && ??

File: vcert/connection_tpp.py [hunk 8]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+                request.id = data['CertificateDN']
+            log.debug("Certificate successfully requested with request id %s." % request.id)
+            return True
+        else:
+            log.error("Request status is not %s. %s." % HTTPStatus.OK, status)
+            raise CertificateRequestError
+
+    @staticmethod
+    def _parse_zone_config_to_policy(data):
+        # todo: parse over values to regexps (dont forget tests!)
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:259]: address todo now ??  as these are lot of if, else in here

File: vcert/connection_tpp.py [hunk 9]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+
     def import_cert(self, request):
         raise NotImplementedError
 
+    @staticmethod
+    def _get_policy_dn(zone):
+        if zone is None:
+            log.error("Bad zone: %s" % zone)
+            raise ClientBadData
+        if re.match(r"^\\\\VED\\\\Policy", zone):
```
Comment (warrior-abhijit) [inline vcert/connection_tpp.py:322]: regex match API would be lot better here and will remove lot of duplicate code below w.r.t regex match

File: vcert/connection_tpp_token.py [hunk 10]
```diff
@@ -128,6 +149,224 @@ def auth(self):
...
+            if request_data.get("CASpecificAttributes"):
+                request_data["CASpecificAttributes"].append(ca_origin)
+            else:
+                request_data["CASpecificAttributes"] = [ca_origin]
+        status, data = self._post(URLS.CERTIFICATE_REQUESTS, data=request_data)
+        if status == HTTPStatus.OK:
+            request.id = data['CertificateDN']
+            log.debug("Certificate sucessfully requested with request id %s." % request.id)
+            return True
+        else:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp_token.py:172]: redundant else

File: vcert/connection_tpp_token.py [hunk 11]
```diff
@@ -128,6 +149,224 @@ def auth(self):
...
+            log.error("Request status is not %s. %s." % HTTPStatus.OK, status)
+            raise CertificateRequestError
+
+    def retrieve_cert(self, certificate_request):
+        log.debug("Getting certificate status for id %s" % certificate_request.id)
+
+        retrive_request = dict(CertificateDN=certificate_request.id, Format="base64", IncludeChain='true')
+
+        if certificate_request.chain_option == "last":
```
Comment (warrior-abhijit) [inline vcert/connection_tpp_token.py:181]: switch case may be here as well ?

File: vcert/connection_tpp_token.py [hunk 12]
```diff
@@ -128,6 +149,224 @@ def auth(self):
...
+
+        status, data = self._post(URLS.CERTIFICATE_RETRIEVE, data=retrive_request)
+        if status == HTTPStatus.OK:
+            pem64 = data['CertificateData']
+            pem = base64.b64decode(pem64)
+            return parse_pem(pem.decode(), certificate_request.chain_option)
+        elif status == HTTPStatus.ACCEPTED:
+            log.debug(data['Status'])
+            return None
+        else:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp_token.py:201]: redundant else

File: vcert/connection_tpp_token.py [hunk 13]
```diff
@@ -128,6 +149,224 @@ def auth(self):
...
+        if reuse_key:
+            log.debug("Trying to renew certificate %s" % request.id)
+            status, data = self._post(URLS.CERTIFICATE_RENEW, data={"CertificateDN": request.id})
+            if not data['Success']:
+                raise CertificateRenewError
+            return
+        cert = self.retrieve_cert(request)
+        cert = x509.load_pem_x509_certificate(cert.cert.encode(), default_backend())
+        for a in cert.subject:
+            if a.oid == x509.NameOID.COMMON_NAME:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp_token.py:242]: switch case again here could be better?

File: vcert/connection_tpp_token.py [hunk 14]
```diff
@@ -128,6 +149,224 @@ def auth(self):
...
+        if not request.csr:
+            request.build_csr()
+        status, data = self._post(URLS.CERTIFICATE_RENEW,
+                                  data={"CertificateDN": request.id, "PKCS10": request.csr})
+        if status == HTTPStatus.OK:
+            if "CertificateDN" in data:
+                request.id = data['CertificateDN']
+            log.debug("Certificate successfully requested with request id %s." % request.id)
+            return True
+        else:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp_token.py:273]: redundant else ?

File: vcert/connection_tpp_token.py [hunk 15]
```diff
@@ -128,6 +149,224 @@ def auth(self):
...
+        if cert.signature_algorithm_oid in (algos.ECDSA_WITH_SHA1, algos.ECDSA_WITH_SHA224, algos.ECDSA_WITH_SHA256,
+                                            algos.ECDSA_WITH_SHA384, algos.ECDSA_WITH_SHA512):
+            request.key_type = (KeyType.ECDSA, KeyType.ALLOWED_CURVES[0])
+        else:
+            request.key_type = KeyType(KeyType.RSA, 2048)  # todo: make parsing key size
+        if not request.csr:
+            request.build_csr()
+        status, data = self._post(URLS.CERTIFICATE_RENEW,
+                                  data={"CertificateDN": request.id, "PKCS10": request.csr})
+        if status == HTTPStatus.OK:
```
Comment (warrior-abhijit) [inline vcert/connection_tpp_token.py:268]: combine 2 ifs into 1

File: vcert/connection_tpp.py (hunk 3 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:181]: removed

File: vcert/connection_tpp.py (hunk 4 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:204]: removed

File: vcert/connection_tpp.py (hunk 2 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:152]: removed

File: tests/test_e2e.py (hunk 1 above)
Comment (rvelaVenafi) [reply tests/test_e2e.py:41]: wildcard import replaced by individual imports

File: vcert/connection_tpp.py (hunk 5 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:222]: native switch-case statement doesn't exist in Python. We can certainly improve the block code with a dictionary, but that change is out of the scope of this PR.

File: vcert/connection_tpp.py (hunk 6 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:253]: removed

File: vcert/connection_tpp_token.py (hunk 11 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp_token.py:181]: no switch in python :(

File: vcert/connection_tpp_token.py (hunk 10 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp_token.py:172]: removed

File: vcert/connection_tpp_token.py (hunk 12 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp_token.py:201]: removed

File: vcert/connection_tpp_token.py (hunk 13 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp_token.py:242]: no switch-case in python :(

File: vcert/connection_tpp_token.py (hunk 14 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp_token.py:273]: removed

File: vcert/connection_tpp.py (hunk 7 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:248]: the CertificateDN value is not necessary to return True, only the status 200.

**Review by warrior-abhijit — APPROVED** [review]

File: vcert/connection_tpp_token.py [hunk 16]
```diff
@@ -128,6 +149,224 @@ def auth(self):
...
+            elif a.oid == x509.NameOID.LOCALITY_NAME:
+                request.locality = a.value
+            elif a.oid == x509.NameOID.STATE_OR_PROVINCE_NAME:
+                request.province = a.value
+            elif a.oid == x509.NameOID.ORGANIZATION_NAME:
+                request.organization = a.value
+            elif a.oid == x509.NameOID.ORGANIZATIONAL_UNIT_NAME:
+                request.organizational_unit = a.value
+        for e in cert.extensions:
+            if e.oid == x509.OID_SUBJECT_ALTERNATIVE_NAME:
```
Comment (LanceGardiner-VEN) [inline vcert/connection_tpp_token.py:255]: Should URI option be included in here?

**Review by LanceGardiner-VEN — APPROVED** [review]

File: vcert/connection_tpp_token.py (hunk 16 above)
Comment (tr1ck3r) [reply vcert/connection_tpp_token.py:255]: @LanceGardiner-VEN Eventually, yes, but that's not related to the token auth updates.  The stated requirement for VCert was to support DNS, IP Address, and Email SAN types.  Support for URI and UPN SAN types was recently added to the VCert-Go project based on need from the Venafi partner ecosystem.  If you're interested in contributing the URI SAN support to VCert-Python we'd welcome it 😃

**Review by LanceGardiner-VEN — APPROVED** [review]

File: vcert/connection_tpp.py [hunk 17]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
+            log.debug("Certificate successfully requested with request id %s." % request.id)
+            return True
+        else:
+            log.error("Request status is not %s. %s." % HTTPStatus.OK, status)
+            raise CertificateRequestError
+
+    @staticmethod
+    def _parse_zone_config_to_policy(data):
+        # todo: parse over values to regexps (dont forget tests!)
+        p = data["Policy"]
```
Comment (HELGAHR) [inline vcert/connection_tpp.py:260]: How safe is it in this method to assume that these dictionary keys resolve? I'm new to this code, but I usually think thrice before trying to access a node in the dictionary without .get().

File: vcert/connection_tpp.py [hunk 18]
```diff
@@ -109,11 +129,220 @@ def auth(self):
...
 
+    @staticmethod
+    def _get_policy_dn(zone):
+        if zone is None:
+            log.error("Bad zone: %s" % zone)
+            raise ClientBadData
+        if re.match(r"^\\\\VED\\\\Policy", zone):
+            return zone
+        else:
+            if re.match(r"^\\\\", zone):
```
Comment (HELGAHR) [inline vcert/connection_tpp.py:325]: Just a tidbit of input: Python string objects have a .startswith() method that's easier to read than a regex, although a regex works fine.

File: vcert/connection_tpp_token.py [hunk 19]
```diff
@@ -18,77 +18,107 @@
...
+class TPPTokenConnection(CommonConnection):
     def __init__(self, url, user=None, password=None, access_token=None, refresh_token=None, http_request_kwargs=None):
-        super().__init__(http_request_kwargs=http_request_kwargs)
+        """
+        :param str url:
+        :param str user:
+        :param str password:
+        :param str access_token:
+        :param str refresh_token:
+        :param dict[str,Any] http_request_kwargs:
```
Comment (HELGAHR) [inline vcert/connection_tpp_token.py:74]: No use in having a doc string if the parameters aren't described, IMO.

File: vcert/connection_tpp_token.py [hunk 20]
```diff
@@ -18,77 +18,107 @@
...
+        if http_request_kwargs is None:
+            http_request_kwargs = {"timeout": 180}
+        elif "timeout" not in http_request_kwargs:
+            http_request_kwargs["timeout"] = 180
+        self._http_request_kwargs = http_request_kwargs or {}
+
+    def __setattr__(self, key, value):
+        if key == "_base_url":
+            value = self._normalize_and_verify_base_url(value)
+        self.__dict__[key] = value
```
Comment (HELGAHR) [inline vcert/connection_tpp_token.py:88]: There is a Python wrapper called @property. Its purpose is to define a getter, setter, and/or deleter method to an instance or class variable. It's nicer for refactoring, mostly. Worth noting if you're interested.

**Review by HELGAHR — APPROVED** [review]

File: vcert/connection_tpp.py [hunk 21]
```diff
@@ -17,46 +17,66 @@
...
-        self.urls[URLS.CERTIFICATE_REVOKE] = URLS.CERTIFICATE_REVOKE
-        self.urls[URLS.CERTIFICATE_RENEW] = URLS.CERTIFICATE_RENEW
-        self.urls[URLS.CERTIFICATE_SEARCH] = URLS.CERTIFICATE_SEARCH
-        self.urls[URLS.CERTIFICATE_IMPORT] = URLS.CERTIFICATE_IMPORT
-        self.urls[URLS.ZONE_CONFIG] = URLS.ZONE_CONFIG
-        self.urls[URLS.CONFIG_READ_DN] = URLS.CONFIG_READ_DN
+    def __setattr__(self, key, value):
+        if key == "_base_url":
+            value = self._normalize_and_verify_base_url(value)
+        self.__dict__[key] = value
```
[the remark this answers is not in the document]
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:76]: Thanks for this suggestion @HELGAHR , we will look into it.

File: vcert/connection_tpp.py (hunk 17 above)
Comment (rvelaVenafi) [reply vcert/connection_tpp.py:260]: Thanks for this suggestion as well. Although the code itself was not added in this PR (it was present before), it is nice to have potentially dangerous code design choices pinpointed

**Pull request merged**